many: merge snap and snapd binaries (#17208)
snapcore/snapd: ARCHITECTURE.md

In both cases, execution starts within the `snap run` command provided with the application (via the symlink) or the service (provided explicitly) reference information.

Both the daemon and the `snap` CLI support *re-exec*: on startup they compare their own version against the `snapd` or `core` snap installed on the system. If the snap is newer, execution restarts from the binary inside that snap — ensuring users always run the most up-to-date snapd without requiring a system package update.

Because `/usr/bin/snap` is a symlink to the `snapd` binary and dispatch is based on `argv[0]`, re-exec into the snap preserves the original `argv[0]` so the same dispatch (CLI vs daemon) applies in the new process.

On a high level, execution of a snap application is carried out in the following manner:

```

>       /snap/bin/<app> or `snap run <snap.app>`

Reviewer: ```suggestion
      /snap/bin/<snap.app> or `snap run <snap.app>`
```

                |

>           (re-exec into snapd snap if newer)

Reviewer: ```suggestion
          (re-exec into snapd/core snap if newer)
```

Author: I don't think we should refence core at this stage, as it's almost always older.

                |
              exec(2)
                |
                v

>           snap run <snap.app>   [snap CLI, argv[0]="snap" or app-name]

Reviewer: Perhaps this step should clarify it is eventually executing snapd binary with argv[0] as "snap" or "snap.app"
```suggestion
          snapd [snap CLI with argv[0]="snap" or "snap.app"]
```

Author: thought it's technically snap CLI that just happens to be implemented in the snapd binary
